cpp/src/arrow/dataset/file_base.cc

```diff
@@ -143,97 +150,235 @@ FragmentIterator FileSystemDataset::GetFragmentsImpl(
+    ARROW_ASSIGN_OR_RAISE(auto scan_task_it,
+                          Scanner(fragment, std::move(options), context).Scan());
+    for (auto maybe_scan_task : scan_task_it) {
+      ARROW_ASSIGN_OR_RAISE(auto scan_task, maybe_scan_task);
+      scan_tasks.push_back(std::move(scan_task));
+      fragment_for_task.push_back(fragment.get());
+    }
+  }
 
-    // make a record batch reader which yields from a fragment
-    ARROW_ASSIGN_OR_RAISE(task->batches, FragmentRecordBatchReader::Make(
-                                             std::move(fragment), schema, scan_context));
-    task_group->Append([task] { return task->Execute(); });
+  // WriteQueues are stored in this deque until writing is completed and are otherwise
+  // referenced by non-owning pointers.
+  std::deque<WriteQueue> queues_storage;
+
+  // Store a mapping from partitions (represened by their formatted partition expressions)
+  // to a WriteQueue which flushes batches into that partition's output file. In principle
+  // any thread could produce a batch for any partition, so each task alternates between
+  // pushing batches and flushing them to disk.
+  util::Mutex queues_mutex;
+  WriteQueue::Set queues;
+
+  auto fragment_for_task_it = fragment_for_task.begin();
+  for (const auto& scan_task : scan_tasks) {
+    const Fragment* fragment = *fragment_for_task_it++;
+
+    task_group->Append([&, scan_task, fragment] {
+      ARROW_ASSIGN_OR_RAISE(auto batches, scan_task->Execute());
+
+      for (auto maybe_batch : batches) {
+        ARROW_ASSIGN_OR_RAISE(auto batch, maybe_batch);
+        ARROW_ASSIGN_OR_RAISE(auto groups, write_options.partitioning->Partition(batch));
+        batch.reset();  // drop to hopefully conserve memory
+
+        std::unordered_set<WriteQueue*> need_flushed;
+        for (size_t i = 0; i < groups.batches.size(); ++i) {
+          AndExpression partition_expression(std::move(groups.expressions[i]),
+                                             fragment->partition_expression());
+          auto batch = std::move(groups.batches[i]);
+
+          ARROW_ASSIGN_OR_RAISE(auto part,
+                                write_options.partitioning->Format(partition_expression));
+
+          util::Mutex::Guard wait_for_opened_writer_lock;
+
+          WriteQueue* queue;
+          {
+            // lookup the queue to which batch should be appended
+            auto queues_lock = queues_mutex.Lock();
+
+            queue = internal::GetOrInsertGenerated(
+                        &queues, std::move(part),
+                        [&](const std::string& emplaced_part) {
+                          // lookup in `queues` also failed,
+                          // generate a new WriteQueue
+                          size_t queue_index = queues.size() - 1;
+
+                          queues_storage.emplace_back(emplaced_part, queue_index,
+                                                      &wait_for_opened_writer_lock);
```

Comment:
This unorthodox thing with the `wait_for_opened_writer_lock` looks like it's because of two-phase initialization...?

Let me suggest something else: make a `WriteQueuePromise` class:
```c++
class WriteQueuePromise {
 public:
  WriteQueuePromise(std::string part, size_t queue_index,
      std::shared_ptr<Schema> schema, const FileSystemDatasetWriteOptions& options)
    : queue_(part, queue_index), schema_(schema), options_(options) {}

  Result<WriteQueue>* operator()() {
    std::call_once(initialized_, [this]() { InitQueue(); });
    RETURN_NOT_OK(status_);
    return &queue_;
  }

 protected:
  void InitQueue() {
    status_ = queue_->OpenWriter(options_, schema_);
  }

  std::once_flag initialized_;
  WriteQueue queue_;
  std::shared_ptr<Schema> schema_;
  const FileSystemDatasetWriteOptions& options_;
  Status status_;
};
```

Also, make `queues` a `std::map<std::string, WriteQueuePromise>`.
Then do something like:
```c++
WriteQueuePromise* queue_promise;
{
  auto queues_lock = queues_mutex.Lock();
  queue_promise = &queues.emplace(part, part, queue_index, batch->schema(), write_options)->second;
}
queue = (*queue_promise)();
```

Reply:
It could also be a `DeferredInit` facility:
```c++
template <typename T>
class DeferredInit<T> {
 public:
  template <typename... Args>
  DeferredInit(Args&&... args)
    : value_(std::forward<Args>(args)...) {}

  Result<T*> operator()() {
    std::call_once(initialized_, [this]() { value_.Init(); });
    RETURN_NOT_OK(status_);
    return &value_;
  }

 protected:
  T value_;
  Status status_;
  std::once_flag initialized_;
};
```

Reply:
If I'm not mistaken, with this approach you don't need the embedded lock guard in `WriteQueue` anymore. Does that make sense?

Reply:
> Let me suggest something else: make a `WriteQueuePromise` class

That sounds doable and would certainly help with encapsulation of the locks.

> If I'm not mistaken, with this approach you don't need the embedded lock guard in WriteQueue anymore. Does that make sense?

`call_once` does guarantee that side effects of the first invocation are observed by all subsequent invocations, so the FileWriter would indeed be initialized for all threads. However we'd still need a `writer_lock` since `FileWriter::Write` is not thread safe

Reply:
Ah... does the `writer_lock` block inside the thread pool?

Reply:
Ah, you're using `TryLock`, nevermind :-)

Reply:
Instead of WriteQueuePromise, I've refactored to initialize the writer in the first call to Flush(), so now all mutexes, locks, and init are internal to WriteQueue.